IOS Sample Update

bboothe-branch opened the pull request.

bboothe-branch committed:
Initial Commit

bboothe-branch committed:
Updated Commit

gdeluna-branch commented on ios_sample_test_app/ios_sample_test_app/AppDelegate.swift:
```diff
@@ -0,0 +1,67 @@
+        Branch.getInstance().setIdentity("your_user_id")
+
+        // logout
+        Branch.getInstance().logout() // or .logoutWithCallback() to customize further
```
Is the logout immediately after setting identity intentional?

bboothe-branch replied:
I had place it there just following [the Branch Docs here.](https://help.branch.io/developers-hub/docs/ios-basic-integration#track-users). I can comment it out or remove it though! Thoughts?

gdeluna-branch replied:
Depends on your intent. The docs have it like that, I presume, to demonstrate the calls, not a real implementation. The identity is set on the session and any calls thenceforth will have that id. 

But, `logout` removes that a few milliseconds later. Maybe you want to call these functions behind buttons?

bboothe-branch replied:
Ok that makes sense. I removed it from AppDelegate and placed it behind the buttons that send events.

bboothe-branch replied:
Thank you

gdeluna-branch commented on ios_sample_test_app/ios_sample_test_app/AppDelegate.swift:
```diff
@@ -0,0 +1,67 @@
+            
+            guard let options = data["nav_to"] as? String else { return }
+            switch options {
+                case "color_block_page":
```
Maybe a little nit but why not an `if`, if you only have one case?

bboothe-branch replied:
Added the other case for going to the Read Deep Link Page

gdeluna-branch commented:
Please add a small description of capabilities, limitations, and some screenshots. ty!

gdeluna-branch commented on ios_sample_test_app/.DS_Store:
This file should be gitignored

gdeluna-branch commented on ios_sample_test_app/ios_sample_test_app/Resources/Fonts/.DS_Store:
same for every ds store file

bboothe-branch replied:
Deleted All DS_Store files

bboothe-branch committed:
DS_Store Files Deleted

bboothe-branch commented:
### Capabilities

**_Home Page:_** 
- App Creates a Branch Link on App Load at the top of the screen
- Send Branch Purchase Event, Add To Cart Event, and Change Background Color Event using the Buttons
- Create a QR Code / Share Branch Link with Navigation Buttons
- Share button creates a link that modifies the color block page
- QR Code button dims screen. Click anywhere on screen outside of Nav Bar to hide QR Code screen.
- Change Background Color button changes the background color along with sending the custom event.

_**Color Block Page:**_
- Color block changes based on "color_block_key" parameter
- Options for "color_block_key" parameter are "red", "green", and "blue".
- If one of the above 3 colors aren't selected, color block defaults to white

_**Read Deep Link Page:**_
- Collects session parameters and stores them
- If a Branch link is clicked to open the app, session parameters are modified to match link parameters.
- Collects install parameters on download and stores them
- Page is scrollable in case parameters are too long for the screen

### Limitations
- Code changes are needed to modify link data on QR Code, Share Sheet, and Generated Branch Link
- All constraints were programmatically created so needs someone that understands those to modify UI
- Install Parameters are difficult to modify
- Routing function uses a method that is technically deprecated

### Screenshots
![IOS App Home Page](https://user-images.githubusercontent.com/104782407/235575905-3be05064-8704-420d-b10b-849e3865b488.png)
![IOS App QR Code](https://user-images.githubusercontent.com/104782407/235575907-a0e08cce-681d-4533-b853-d2b7b96103b6.png)
![IOS App Read Deep Link Page](https://user-images.githubusercontent.com/104782407/235575910-272396b9-3eab-4598-ad3a-06473f2a704a.png)
![IOS App Color Block Green](https://user-images.githubusercontent.com/104782407/235575912-cb650fbb-bc0e-453a-b9a3-6a1bc1ebad02.png)
![IOS App Color Block Neutral](https://user-images.githubusercontent.com/104782407/235575913-1caecb7e-0678-4808-965b-3ea967607444.png)

bboothe-branch commented:
> Please add a small description of capabilities, limitations, and some screenshots. ty!

@gdeluna-branch Added a comment with capabilities, limitations and screenshots

bboothe-branch committed:
Duplicated Gitignore files

nsingh-branch commented on ios_sample_test_app/Podfile:
Since you're adding Branch via SPM, you can remove Cocoapods from the project

bboothe-branch replied:
Followed the below info with pod deintegrate and pod clean to remove Cocoapods from project

https://stackoverflow.com/questions/16427421/how-to-remove-cocoapods-from-a-project

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+        super.viewDidLoad()
+        self.navigationController?.navigationBar.largeTitleTextAttributes = [NSAttributedString.Key.font: UIFont(name: "Raleway-Bold", size: 30.0)!]
+        self.navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.font: UIFont(name: "Raleway-SemiBold", size: 18.0)!]
+        view.backgroundColor = .white
```
I would recommend setting the background color to .systemBackground so it can handle dark mode. You may have to do this on the other pages as well.

bboothe-branch replied:
Completed

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+        ]
+        event.logEvent()
+        // logout
+        Branch.getInstance().logout() // or .logoutWithCallback() to customize further
```
I noticed that logout() is called after each event, but this might not be necessary.

bboothe-branch replied:
removed logout() from events as it is unnecessary

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+    // ---------- Branch SDK Functions ----------
+    
+    func createBranchLink() {
+        let buo = BranchUniversalObject.init(canonicalIdentifier: "context/12345")
+        buo.title = "My Content Title"
+        buo.contentDescription = "My Content Description"
+        buo.imageUrl = "https://lorempixel.com/400/400"
+        buo.publiclyIndex = true
+        buo.locallyIndex = true
+        buo.contentMetadata.customMetadata["key1"] = "value1"
+        
+        let lp: BranchLinkProperties = BranchLinkProperties()
+        lp.channel = "facebook"
+        lp.feature = "sharing"
+        lp.campaign = "content 123 launch"
+        lp.stage = "new user"
+        lp.tags = ["one", "two", "three"]
```
I would consolidate the code to create the BUO and LP objects since its repeated a few times. It could be made into a function or just defined once and reused.

bboothe-branch replied:
Consolidated buo and lp into functions and modified code in Branch functions to reuse them.

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+    }
+    
+    @objc func shareBranchLink() {
+        let buo = BranchUniversalObject.init(canonicalIdentifier: "context/23456")
```
```suggestion
        let buo = BranchUniversalObject(canonicalIdentifier: "context/23456")
```
You can initialize without the init as well.

bboothe-branch replied:
Removed .init from BUO events

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+        super.viewDidLoad()
+        self.navigationController?.navigationBar.largeTitleTextAttributes = [NSAttributedString.Key.font: UIFont(name: "Raleway-Bold", size: 30.0)!]
+        self.navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.font: UIFont(name: "Raleway-SemiBold", size: 18.0)!]
+        view.backgroundColor = .white
```
```suggestion
        view.backgroundColor = .systemBackground
```

bboothe-branch replied:
resolved in addition to above

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+        qrCode.margin = 0
+        qrCode.imageFormat = .JPEG
+        
+        qrCode.getAsImage(buo, linkProperties: lp) { qrCodeImage, error in
```
To be safe, i'd recommend checking for an error first, just in case the method was to fail.

bboothe-branch replied:
Used [this example](https://stackoverflow.com/questions/41717170/using-guard-to-check-for-nil-without-implicitly-unwrapping) to find the best way to do this and modified to the code below.

```
 qrCode.getAsImage(buo, linkProperties: lp) { qrCodeImage, error in
            // Checking for error prior to initializing QR Code
            if (error != nil) {
                print(error as Any)
                return
            }
            
            // Do something with your QR code here...
            DispatchQueue.main.async {
                self.qrCode.backgroundColor = UIColor(patternImage: qrCodeImage!)
            }
        }
```

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+
+        NSLayoutConstraint.activate([
```
For the constraints, I'd recommend trying to stick with leading and trailing constraints instead of width and centerX. Otherwise, it may look wrong on certain device sizes or lead to redundant constraints.

bboothe-branch replied:
Modified all width and centerX constraints to leading & trailing constraints

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+
+// Enables setting colors using hex codes
+
+extension UIColor {
```
If possible, I would just hardcode the RGB values or create a custom color asset for the hex colors the app is using instead of including this code. I recommend this site which makes it really easy www.uicolor.io.

bboothe-branch replied:
Hardcoded all values using that site. Thanks for that! Would have made the original coloring a lot easier

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+    }()
+    
+    private lazy var addToCartButton: UIButton = {
+       let button = UIButton()
+        button.backgroundColor = UIColor(rgb: 0x0074DF)
+        button.setTitle("Add To Cart", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.titleLabel?.font = UIFont(name: "Raleway-SemiBold", size: 20)
+        button.layer.cornerRadius = 27.5
+        button.layer.cornerCurve = .continuous
+        button.addTarget(self, action: #selector(sendAddToCartEvent(_:)), for: .touchUpInside)
+        button.translatesAutoresizingMaskIntoConstraints = false
+        return button
+    }()
+    
+    private lazy var colorBlockPageButton: UIButton = {
+        let button = UIButton()
+        button.backgroundColor = UIColor(rgb: 0x0074DF)
+        button.setTitle("Color Block Page", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.titleLabel?.font = UIFont(name: "Raleway-SemiBold", size: 20)
+        button.layer.cornerRadius = 27.5
+        button.layer.cornerCurve = .continuous
+        button.addTarget(self, action: #selector(openColorBlockPage(_:)), for: .touchUpInside)
+        button.translatesAutoresizingMaskIntoConstraints = false
+        return button
+    }()
+    
+    private lazy var readDeepLinkButton: UIButton = {
+       let button = UIButton()
+        button.backgroundColor = UIColor(rgb: 0x0074DF)
+        button.setTitle("Read Link Data", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.titleLabel?.font = UIFont(name: "Raleway-SemiBold", size: 20)
+        button.layer.cornerRadius = 27.5
+        button.layer.cornerCurve = .continuous
+        button.addTarget(self, action: #selector(openReadDeepLinkPage(_:)), for: .touchUpInside)
+        button.translatesAutoresizingMaskIntoConstraints = false
+        return button
+    }()
```
Since these buttons share a lot of the same properties, you could create a simple function returns a UIButton and takes a title string and action selector.

bboothe-branch replied:
Created function that takes in the button title and sets both title and action based on the title

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/Info.plist:
Lets update the Branch Key to a new one that can be shared between all of the test apps.

bboothe-branch replied:
Agreed. Will put that dashboard together. I'll have to update the Android one as well

bboothe-branch replied:
Will create a new PR for Android, but created a new dashboard and added Live and Test Keys

Branch App ID: 1148693269161464775

Also added App Tracking Transparency Prompt for Device Level data to be sent to dashboard

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,488 @@
+    
+    // ---------- UILabels and UIImages ----------
+    
+    private let generatedBranchLink: UILabel = {
+        let label = UILabel()
+        label.textColor = UIColor(rgb: 0x1F2852)
+        label.numberOfLines = 1
+        label.textAlignment = .center
+        label.font = UIFont(name: "Raleway-Medium", size: 14)
+        label.layer.borderColor = UIColor(rgb: 0xBCC0D4).cgColor
+        label.layer.borderWidth = 1.5
+        label.layer.cornerRadius = 5
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
```
Let's update this from a UILabel to a UITextView so the link can be copied and tapped.

bboothe-branch replied:
Modified to UITextView

nsingh-branch reviewed (changes requested):
Overall, great job. The code is very readable and everything's working well. I just had some comments on how some of the duplicate code can be reduced, some visual bugs with different settings/screen sizes, and a few other small things. 

Let me know if I can clarify any of the comments as well.

bboothe-branch committed:
Modifications After Nipun's Review

Let me know if it all looks good

bboothe-branch committed:
d

bboothe-branch committed:
Fixed accidental commit

bboothe-branch committed:
Added README.md

nsingh-branch commented on ios_sample_test_app/ios_sample_test_app/ViewController.swift:
```diff
@@ -0,0 +1,411 @@
+            generatedBranchLink.heightAnchor.constraint(equalToConstant: 33),
+
+            branchBadgeImageView.topAnchor.constraint(equalTo: generatedBranchLink.bottomAnchor, constant: 10),
+            branchBadgeImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 20),
```
```suggestion
                        branchBadgeImageView.centerXAnchor.constraint(equalTo: contentView.centerXAnchor),
```

Since there is a width and leader constraint, the image view is a bit off-centered. Updating the leading constraint to a centerX constraint can fix this,

bboothe-branch replied:
Added center X anchor constraints

nsingh-branch reviewed (changes requested):
Last couple things would be to fix some of the dark blue text being hard to see when viewed in dark mode, specifically the main link text field and the labels on the inner pages. You can use `traitCollection.userInterfaceStyle` to detect whether the user is in light or dark mode and change the text color based on that.

bboothe-branch committed:
Dark Mode Updates

Added updates to Dark mode and centered the Branch badge.

Also slightly modified the text in color block + deep link page to be larger for visibility.

bboothe-branch committed:
Added Description to Dark Mode check

nsingh-branch commented:
UI looks good now, there are just a few small things left. It appears that deep linking isn't working properly, which may be a configuration issue. Additionally, the QR code API still doesn't appear to be enabled for this key, but I can reach out to Core to have that fixed. I'd also recommend updating the Display Name to something like "Branch Sample" and the Bundle Identifier to "io.branch.branchsample-ios" just for consistency.

bboothe-branch committed:
Updated Display Name and bundle ID

bboothe-branch committed:
Updated Bundle ID on Signing and Capabilities

nsingh-branch committed:
added display name

nsingh-branch reviewed (approved):
@bboothe-branch  LGTM! I just added a display name so it looks good to merge now.
